File: tensorflow_py/tf_executor.py
import argparse
import tensorflow as tf
import numpy as np
import time

CONST_VALUE = .00000182
GLOBAL_DATA_FORMAT = "NCHW"

class PaddingType:
  SAME = 0
  VALID = 1

class ActiMode:
  AC_MODE_NONE = 0
  AC_MODE_SIGMOID = 1
  AC_MODE_RELU = 2
  AC_MODE_TANH = 3

class OpType:
  OP_NOOP = 0
  OP_ANY = 1
  OP_CONV2D = 2
  OP_LINEAR = 3
  OP_POOL2D_MAX = 4
  OP_POOL2D_AVG = 5
  OP_RELU = 6
  OP_SIGMOID = 7
  OP_BATCHNORM = 8
  OP_CONCAT = 9
  OP_SPLIT = 10
  # RNN operators
  OP_EW_ADD = 11
  OP_EW_MUL = 12
  OP_MATMUL = 13

def get_padding_string(padding_type):
  if (padding_type == PaddingType.SAME):
    return "SAME"
  elif (padding_type == PaddingType.VALID):
    return "VALID"
  else:
    print("Unknown padding string")
    assert(0)

def split_string_ints(string, delim):
  ints = []
  splits = string.split(delim)
  for s in splits:
    ints.append(int(s))
  return ints

def split_string_int_pairs(string, delim1, delim2):
  pairs = []
  splits = string.split(delim1)
  for s in splits:
    pair_split = s.split(delim2)
    pairs.append((int(pair_split[0]), int(pair_split[1])))
  return pairs

def make_conv2d_with_bias(input_tensor, filter_shape, strides, padding, bias_dim, add_relu, name):
  weights_name = name + "_weights"
  bias_name = name + "_bias"
  conv_name = name + "_conv2d"
  bias_add_name = name + "_bias_add"

  weights = tf.constant(CONST_VALUE, shape=filter_shape, name=weights_name)
  bias = tf.constant(CONST_VALUE, shape=[bias_dim], name=bias_name)
  conv2d = tf.nn.conv2d(input_tensor, weights, strides, get_padding_string(padding), data_format=GLOBAL_DATA_FORMAT, name=conv_name)
  bias_add = tf.nn.bias_add(conv2d, bias, data_format=GLOBAL_DATA_FORMAT, name=bias_add_name)

  if (add_relu):
    relu_name = name + "_relu"
    relu = tf.nn.relu(bias_add, name=relu_name)
    return relu
  else:
    return bias_add

def create_input(line, operator_map):
  dims = split_string_ints(line, ',')
  input_shape = []
  for i in xrange(0, len(dims)):
    if (dims[i] > 0):
      input_shape.append(dims[i])
  input_placeholder = tf.placeholder(tf.float32, shape=input_shape)
  operator_map[(0,0)] = input_placeholder
  return input_shape

def parse_operator(line1, line2, line3, line4, operator_map, graph_outputs):
  guid = int(line1) 
  op_type = int(line2)
  deps = split_string_int_pairs(line3, ',', ':')
  for dep in deps:
    if dep in graph_outputs:
      graph_outputs.remove(dep)
  if (op_type == OpType.OP_CONV2D):
    params = split_string_ints(line4, ',')
    filter_shape = [params[5], params[6], params[1], params[4]]
    strides = [1, 1, params[7], params[8]]
    if (params[9] > 0 or params[10] > 0):
      padding = PaddingType.SAME
    else:
      padding = PaddingType.VALID
    name = "conv2d_" + str(guid)
    conv = make_conv2d_with_bias(operator_map[deps[0]], filter_shape, strides, padding, params[4], params[11], name=name)
    operator_map[(guid,0)] = conv
    return [(guid,0)]
  elif (op_type == OpType.OP_POOL2D_MAX or op_type == OpType.OP_POOL2D_AVG):
    params = split_string_ints(line4, ',')
    ksize = [1, 1, params[5], params[6]]
    strides = [1, 1, params[7], params[8]]
    if (params[9] > 0 or params[10] > 0):
      padding = PaddingType.SAME
    else:
      padding = PaddingType.VALID
    created_op = None
    name = None
    if (op_type == OpType.OP_POOL2D_MAX):
      name = "maxpool_" + str(guid)
      max_pool = tf.nn.max_pool(operator_map[deps[0]], ksize, strides, get_padding_string(padding), data_format=GLOBAL_DATA_FORMAT, name=name)
      created_op = max_pool
    else:
      name = "avgpool_" + str(guid)
      avg_pool = tf.nn.avg_pool(operator_map[deps[0]], ksize, strides, get_padding_string(padding), data_format=GLOBAL_DATA_FORMAT, name=name)
      created_op = avg_pool
    if params[10]: # If add relu
      relu_name = name + "_relu"
      relu = tf.nn.relu(created_op, name=relu_name)
      operator_map[(guid,0)] = relu
      return [(guid,0)]
    else:
      operator_map[(guid,0)] = created_op
      return [(guid,0)]
  elif (op_type == OpType.OP_SPLIT):
    params = split_string_ints(line4, ',')
    name = "split_" + str(guid)
    splits = tf.split(operator_map[deps[0]], params, 1, name=name)
    rets = []
    for i in xrange(0, len(splits)):
      operator_map[(guid, i)] = splits[i]
      rets.append((guid, i))
    return rets
  elif (op_type == OpType.OP_CONCAT):
    name = "concat_" + str(guid)
    dep_tensors = []
    for i in xrange(0, len(deps)):
      dep_tensors.append(operator_map[deps[i]])
    concat = tf.concat(dep_tensors, 1, name=name)
    operator_map[(guid,0)] = concat
    return [(guid,0)]
  elif (op_type == OpType.OP_EW_ADD):
    name = "ew_add_" + str(guid)
    ew_add = tf.add(operator_map[deps[0]], operator_map[deps[1]], name=name)
    operator_map[(guid,0)] = ew_add
    return [(guid,0)]
  elif (op_type == OpType.OP_EW_MUL):
    name = "ew_mul_" + str(guid)
    ew_mul = tf.multiply(operator_map[deps[0]], operator_map[deps[1]], name=name)
    operator_map[(guid,0)] = ew_mul
    return [(guid,0)]
  elif (op_type == OpType.OP_RELU):
    name = "relu_" + str(guid)
    relu = tf.nn.relu(operator_map[deps[0]], name=name)
    operator_map[(guid,0)] = relu
    return [(guid,0)]
  elif (op_type == OpType.OP_SIGMOID):
    name = "sigmoid_" + str(guid)
    sigmoid = tf.nn.sigmoid(operator_map[deps[0]], name=name)
    operator_map[(guid,0)] = sigmoid
    return [(guid,0)]
  elif (op_type == OpType.OP_BATCHNORM or op_type == OpType.OP_NOOP):
    operator_map[(guid,0)] = operator_map[deps[0]]
    return [(guid,0)]
  elif (op_type == OpType.OP_MATMUL):
    params = split_string_ints(line4, ',')
    assert(len(params) == 5)
    name = "matmul_" + str(guid)
    reshape_name = name + "_reshape"
    weights_name = name + "_weights"
    bias_name = name + "_bias"
    matmul_name = name + "_matmul"
    bias_add_name = name + "_bias_add"
    weights_shape = [params[2], params[3]]
    bias_dim = params[3]
    weights = tf.constant(CONST_VALUE, shape=weights_shape, name=weights_name)
    bias = tf.constant(CONST_VALUE, shape=[bias_dim], name=bias_name)
    reshape = tf.reshape(operator_map[deps[0]], [params[0]*params[1], params[2]], name=reshape_name)
    matmul = tf.matmul(reshape, weights, name=matmul_name)
    bias_add = tf.nn.bias_add(matmul, bias, name=bias_add_name)
    actimode = params[4]
    if (actimode == ActiMode.AC_MODE_NONE):
      operator_map[(guid,0)] = bias_add
      return [(guid,0)]
    elif (actimode == ActiMode.AC_MODE_SIGMOID):
      name += "_sigmoid"
      sigmoid = tf.nn.sigmoid(bias_add, name=name)
      operator_map[(guid,0)] = sigmoid
      return [(guid,0)]
    elif (actimode == ActiMode.AC_MODE_RELU):
      name += "_relu"
      relu = tf.nn.relu(bias_add, name=name)
      operator_map[(guid,0)] = relu
      return [(guid,0)]
    elif (actimode == ActiMode.AC_MODE_TANH):
      name += "_tanh"
      tanh = tf.nn.tanh(bias_add, name=name)
      operator_map[(guid,0)] = tanh
      return [(guid,0)]
    else:
      print("unknown actimode!!!!")
      assert(0)
  else:
    print("Found unknown opcode")
    assert(0)


parser = argparse.ArgumentParser()
parser.add_argument("--xla", help="Whether to run with TensorFlowXLA optimizations", action="store_true")
parser.add_argument("--model_file", help="The file from which to load the model")
parser.add_argument("--print_tensorboard", help="Name of folder to output the tensorboard information")
parser.add_argument("--iterations", help="How many iterations to average for timing (default 5000)", type=int, default=50)
parser.add_argument("--discard_iter", help="How many iterations to not time during warm up (default 1000)", type=int, default=10)
args = parser.parse_args()

input_shape = []
graph_outputs = set()

with open(args.model_file, 'r') as graph_file:
  # The graph nodes are repesented by 4 lines
  operator_map = {}
  need_input = True
  line1 = graph_file.readline()

  while line1:
    line2 = graph_file.readline()
    line3 = graph_file.readline()
    line4 = graph_file.readline()
    # Cut off the newlines
    line1 = line1[0:-1]
    line2 = line2[0:-1]
    line3 = line3[0:-1]
    line4 = line4[0:-1]

    if (need_input):
      need_input = False
      input_shape = create_input(line4, operator_map)
      graph_outputs.add((0,0))

    recent_outputs = parse_operator(line1, line2, line3, line4, operator_map, graph_outputs)
    for output in recent_outputs:
      graph_outputs.add(output)

    # Using this as the test of if the file is empty
    line1 = graph_file.readline()

if (len(graph_outputs) == 0):
  print("Could not read the graph!!!")
  assert(0)

output_nodes = []
for graph_output in graph_outputs:
  output_nodes.append(operator_map[graph_output])

config = tf.ConfigProto()
if (args.xla):
  config.graph_options.optimizer_options.global_jit_level = tf.OptimizerOptions.ON_1

input_data = np.random.random_sample(input_shape)
with tf.Session(config=config) as sess:
  if (args.print_tensorboard):
    writer = tf.summary.FileWriter(args.print_tensorboard, sess.graph)
  times = []
  for i in range(args.discard_iter + args.iterations):
    t0 = time.time()
    sess.run(output_nodes, {operator_map[(0,0)]: input_data})
    t1 = time.time()
    print(str(t1 - t0) + " seconds")
    times.append(t1 - t0)

  total = 0
  for i in range(args.discard_iter, len(times)):
    total += times[i]
  avg = total / (args.iterations)
  print("Average time of the last " + str(args.iterations) + " iterations: " + str(avg) + " seconds")

File: tensorflow_cpp/tf_executor.cc
// tensorflow/cc/example/example.cc

#include "tensorflow/cc/client/client_session.h"
#include "tensorflow/cc/ops/standard_ops.h"
#include "tensorflow/core/framework/tensor.h"

#include <assert.h>
#include <iostream>
#include <fstream>
#include <string>
#include <time.h>
#include <chrono>
using namespace std;


enum DataFormat {
  NCHW,
  NHWC,
};

enum PaddingType {
  SAME,
  VALID,
};

// This must be kept in sync with the type definitions in the graph code
// for now.
enum OpType {
  OP_NOOP, // 0
  OP_ANY, // 1
  OP_CONV2D, // 2
  OP_LINEAR, // 3
  OP_POOL2D_MAX, // 4
  OP_POOL2D_AVG, // 5
  OP_RELU, // 6
  OP_BATCHNORM, // 7
  OP_CONCAT, // 8
  OP_SPLIT, // 9
  // RNN operators
  OP_EW_ADD, // 10
  OP_EW_MUL, // 11
  OP_MATMUL, // 12
};


std::string get_data_format_string(DataFormat format)
{
  switch (format)
  {
    case NCHW:
    {
      return "NCHW";
    }
    case NHWC:
    {
      return "NHWC";
    }
    default:
    {
      abort();
      return "fuck compiler warnings";
    }
  }
}

std::string get_padding_type_string(PaddingType padding)
{
  switch (padding)
  {
    case SAME:
      return "SAME";
    case VALID:
      return "VALID";
    default:
      assert(0);
      return "fuck compiler warnings";
  }
}

tensorflow::TensorShape create_tensor_shape(std::vector<int> dims)
{
  using namespace tensorflow;
  TensorShape shape;
  for (int i = 0; i < dims.size(); i++)
  {
    shape.AddDim(dims[i]);
  }
  return shape;
}

// Want a "conv2d" maker that takes an input shape, input, weight shape, weights, and then creates the convolution with bias and activation

tensorflow::Output make_conv2d_with_bias(tensorflow::Input input,
                                         std::vector<int> filter_shape,
                                         std::vector<int> strides,
                                         DataFormat format,
                                         PaddingType padding,
                                         int bias_dim,
                                         bool add_relu,
                                         std::string name,
                                         tensorflow::Scope &scope)
{
  const float CONST_VALUE = .00000182;

  // First, let's make sure that things are being used correctly
  assert(input_shape == 4);
  assert(filter_shape.size() == 4);
  assert(strides.size() == 4);
  assert(strides[0] == 1);
  if (format == NCHW)
  {
    assert(strides[1] == 1);
  }
  else
  {
    assert(strides[3] == 1);
  }

  std::string weights_name(name);
  std::string bias_name(name);
  std::string conv_name(name);
  std::string bias_add_name(name);
  weights_name += "_weights";
  bias_name += "_bias";
  conv_name += "_conv2d";
  bias_add_name += "_bias_add";

  auto weights = tensorflow::ops::Const(scope.WithOpName(weights_name.c_str()), CONST_VALUE,
                                        create_tensor_shape(filter_shape));
  if (!scope.ok()) {
    LOG(FATAL) << scope.status().ToString();
    abort();
  }
  auto bias = tensorflow::ops::Const(scope.WithOpName(bias_name.c_str()), CONST_VALUE,
                                     {bias_dim});

  tensorflow::ops::Conv2D::Attrs attrs =
      tensorflow::ops::Conv2D::DataFormat("NCHW");

  auto conv2d = tensorflow::ops::Conv2D(scope.WithOpName(conv_name.c_str()), input,
                            weights, strides, get_padding_type_string(padding).c_str(), attrs);
  if (!scope.ok()) {
    LOG(FATAL) << scope.status().ToString();
    abort();
  }
  tensorflow::ops::BiasAdd::Attrs bias_attrs =
      tensorflow::ops::BiasAdd::DataFormat("NCHW");
  auto bias_add = tensorflow::ops::BiasAdd(scope.WithOpName(bias_add_name.c_str()), conv2d, bias, bias_attrs);
  if (!scope.ok()) {
    LOG(FATAL) << scope.status().ToString();
    abort();
  }

  if (add_relu)
  {
    std::string relu_name(name);
    relu_name += "_relu";
    auto relu_node =
        tensorflow::ops::Relu(scope.WithOpName(relu_name.c_str()), bias_add);
    if (!scope.ok()) {
      LOG(FATAL) << scope.status().ToString();
      abort();
    }
    return relu_node;
  }
  else
  {
    return bias_add;
  }
}

void split_string(const std::string &str,
                  char delimiter,
                  std::vector<std::string> &splits) {
  std::size_t from = 0;
  for (std::size_t i = 0; i < str.size(); ++i) {
    if (str[i] == delimiter) {
      splits.push_back(str.substr(from,i - from));
      from = i + 1;
    }
  }
  if (from < str.size())
  {
    splits.push_back(str.substr(from,str.size() - from));
  }
}

void split_string_ints(const std::string &str,
                       char delimiter,
                       std::vector<int> &splits) {
  std::vector<std::string> str_splits;
  split_string(str, delimiter, str_splits);
  for (unsigned i = 0; i < str_splits.size(); i++)
  {
    splits.push_back(atoi(str_splits[i].c_str()));
  }
}

void split_string_int_pairs(const std::string &str,
                            char delimiter1,
                            char delimiter2,
                            std::vector<std::pair<int,int> > &splits) {
  std::vector<std::string> str_splits;
  split_string(str, delimiter1, str_splits);
  for (unsigned i = 0; i < str_splits.size(); i++)
  {
    std::vector<std::string> second_splits;
    split_string(str_splits[i], delimiter2, second_splits);
    splits.push_back(
        std::make_pair(atoi(second_splits[0].c_str()),
                       atoi(second_splits[1].c_str())));
  }
}

void create_input(std::string &line,
                  tensorflow::Scope &scope,
                  DataFormat data_format,
                  std::map<std::pair<int, int>, tensorflow::Output> &operator_map)
{
  std::vector<int> shape;
  shape.push_back(1);
  split_string_ints(line, ',', shape);
  auto attrs =
      tensorflow::ops::Placeholder::Attrs().Shape(create_tensor_shape(shape));
  tensorflow::Output input = tensorflow::ops::Placeholder(
      scope.WithOpName("input"), tensorflow::DT_FLOAT, attrs);
  operator_map[std::make_pair(0,0)] = input;
}


tensorflow::OutputList parse_operator(std::string &line1,
    std::string &line2,
    std::string &line3,
    std::string &line4,
    tensorflow::Scope &scope,
    DataFormat data_format,
    std::map<std::pair<int, int>, tensorflow::Output> &operator_map)
{
  int guid = atoi(line1.c_str());
  int type = atoi(line2.c_str());

  std::vector<std::pair<int,int> > deps;
  split_string_int_pairs(line3, ',', ':', deps);

  switch(type)
  {
    case OP_CONV2D:
    { 
      std::vector<int> params;
      split_string_ints(line4, ',', params);
      std::vector<int> filter_shape;
      filter_shape.push_back(params[4]);
      filter_shape.push_back(params[5]);
      filter_shape.push_back(params[0]);
      filter_shape.push_back(params[3]);
      std::vector<int> strides;
      strides.push_back(1);
      strides.push_back(1);
      strides.push_back(params[6]);
      strides.push_back(params[7]);

      PaddingType padding;
      if (params[8] > 0 || params[9] > 0)
      {
        padding = SAME;
      }
      else
      {
        padding = VALID;
      }

      std::string name = "conv2d_";
      name += std::to_string(guid);

      tensorflow::Output conv = make_conv2d_with_bias(
          operator_map[deps[0]], filter_shape, strides,
          data_format, padding, params[3], params[10], "conv2d", scope);
      operator_map[std::make_pair(guid, 0)] = conv;
      return {conv};
    }
    case OP_POOL2D_MAX:
    case OP_POOL2D_AVG:
    {
      std::vector<int> params;
      split_string_ints(line4, ',', params);
      std::vector<int> ksize;
      ksize.push_back(1);
      ksize.push_back(1);
      ksize.push_back(params[4]);
      ksize.push_back(params[5]);
      std::vector<int> strides;
      strides.push_back(1);
      strides.push_back(1);
      strides.push_back(params[6]);
      strides.push_back(params[7]);

      PaddingType padding;
      if (params[8] > 0 || params[9] > 0)
      {
        padding = SAME;
      }
      else
      {
        padding = VALID;
      }

      if (type == OP_POOL2D_MAX)
      {
        std::string name = "maxpool_";
        name += std::to_string(guid);
        tensorflow::ops::MaxPool::Attrs attrs =
            tensorflow::ops::MaxPool::DataFormat("NCHW");
        tensorflow::Output max_pool = tensorflow::ops::MaxPool(
            scope.WithOpName(name.c_str()), operator_map[deps[0]],
            ksize, strides, get_padding_type_string(padding).c_str(), attrs);
        operator_map[std::make_pair(guid, 0)] = max_pool;
        if (!scope.ok()) {
          LOG(FATAL) << scope.status().ToString();
          abort();
        }
        return {max_pool};
      }
      else
      {
        std::string name = "avgpool_";
        name += std::to_string(guid);
        tensorflow::ops::AvgPool::Attrs attrs =
            tensorflow::ops::AvgPool::DataFormat("NCHW");
        tensorflow::Output avg_pool = tensorflow::ops::AvgPool(
            scope.WithOpName(name.c_str()), operator_map[deps[0]],
            ksize, strides, get_padding_type_string(padding).c_str(), attrs);
        operator_map[std::make_pair(guid, 0)] = avg_pool;
        if (!scope.ok()) {
          LOG(FATAL) << scope.status().ToString();
          abort();
        }
        return {avg_pool};
      }
    }
    case OP_SPLIT:
    {
      std::vector<int> params;
      split_string_ints(line4, ',', params);
      int num_splits = params.size();

      tensorflow::Tensor split_sizes(tensorflow::DT_INT32,
          tensorflow::TensorShape({4}));

      for (int i = 0; i < num_splits; i++)
      {
        split_sizes.vec<int>()(i) = params[i];
      }

      std::string name = "split_";
      name += std::to_string(guid);
      auto splits = tensorflow::ops::SplitV(
          scope.WithOpName(name.c_str()), operator_map[deps[0]],
          split_sizes, 1 /* split axis */, num_splits);

      for (int i = 0; i < num_splits; i++)
      {
        operator_map[std::make_pair(guid, i)] = splits[i];
      }
      return splits.output;
    }
    case OP_CONCAT:
    {
      std::string name = "concat_";
      name += std::to_string(guid);
      tensorflow::OutputList inputs_as_outputs;
      for (int i = 0; i < deps.size(); i++)
      {
        inputs_as_outputs.push_back(operator_map[deps[i]]);
      }
      tensorflow::Output concat = tensorflow::ops::Concat(
          scope.WithOpName(name.c_str()), inputs_as_outputs, 1 /* split axis */);

      operator_map[std::make_pair(guid, 0)] = concat;
      return {concat};
    }
    case OP_EW_ADD:
    {
      std::string name = "add_";
      name += std::to_string(guid);
      tensorflow::Output add = tensorflow::ops::Add(
          scope.WithOpName(name.c_str()),
          operator_map[deps[0]],
          operator_map[deps[1]]);

      operator_map[std::make_pair(guid, 0)] = add;
      return {add};
    }
    case OP_EW_MUL:
    {
      std::string name = "mul_";
      name += std::to_string(guid);
      tensorflow::Output mul = tensorflow::ops::Multiply(
          scope.WithOpName(name.c_str()),
          operator_map[deps[0]],
          operator_map[deps[1]]);

      operator_map[std::make_pair(guid, 0)] = mul;
      return {mul};
    }
    case OP_RELU:
    {
      std::string name = "relu_";
      name += std::to_string(guid);
      tensorflow::Output relu = tensorflow::ops::Relu(
          scope.WithOpName(name.c_str()),
          operator_map[deps[0]]);

      operator_map[std::make_pair(guid, 0)] = relu;
      return {relu};
    }
    case OP_BATCHNORM:
    case OP_NOOP:
    {
      operator_map[std::make_pair(guid, 0)] = operator_map[deps[0]];
      return {operator_map[deps[0]]};
    }
    case OP_MATMUL: // This doesn't seem to be implemented in run either
    default:
    {
      abort();
      return {operator_map[deps[0]]};
    }
  }
  
}


std::pair<tensorflow::Output, tensorflow::OutputList> buildTFGraphFromFile(
    tensorflow::Scope &scope,
    std::string file_name,
    DataFormat data_format)
{
  tensorflow::OutputList recent_outputs;
  std::map<std::pair<int, int>, tensorflow::Output> operator_map;

  std::string line1;
  std::string line2;
  std::string line3;
  std::string line4;
  ifstream myfile (file_name.c_str());
  if (myfile.is_open())
  {
    // create the placeholder for the input
    bool need_input = true;
    while ( getline(myfile,line1) )
    {
      // The format encodes nodes in 4 lines
      getline(myfile,line2);
      getline(myfile,line3);
      getline(myfile,line4);

      if (need_input)
      {
        need_input = false;
        create_input(line4, scope, data_format, operator_map);
      }

      recent_outputs = parse_operator(line1, line2, line3, line4, scope,
                     data_format, operator_map);
    }
    myfile.close();
  }
  return std::make_pair(operator_map[std::make_pair(0,0)], recent_outputs);
}


void print_timediff(const char* prefix, const struct timespec& start, const 
struct timespec& end)
{
      double milliseconds = (end.tv_nsec - start.tv_nsec) / 1e6 + (end.tv_sec - start.tv_sec) * 1e3;
          printf("%s: %lf milliseconds\n", prefix, milliseconds);
}

int main(int argc, char **argv)
{
  bool use_xla = false;

  for (int i = 1; i < argc; i++)
  {
    if (!strcmp(argv[i],"--xla")) {
      use_xla = true;
      continue;
    }
    fprintf(stderr, "Found unknown option!!\n");
    abort();
  }

  std::string FILE_NAME = "/home/users/twarszaw/tensorflow/tensorflow/cc/graph_transfer/tensorflow_cpp/test_squeeze";
  tensorflow::Scope root = tensorflow::Scope::NewRootScope();
  std::pair<tensorflow::Output, tensorflow::OutputList> input_and_outputs =
      buildTFGraphFromFile(root, FILE_NAME.c_str(), NCHW);
  tensorflow::OutputList outputs = input_and_outputs.second;

  const float CONST_VALUE = .00000182;
  //TODO: read this starting shape from the input
  std::vector<int> starting_shape({1, 3, 222, 222});

  tensorflow::Input::Initializer input_tensor(CONST_VALUE,
                                  create_tensor_shape(starting_shape));

  std::vector<tensorflow::Tensor> output_tensors;

  tensorflow::SessionOptions sess_options;
  if (use_xla || true)
  {
    sess_options.config.mutable_graph_options()->mutable_optimizer_options()->set_global_jit_level(tensorflow::OptimizerOptions::ON_2);
  }
  fprintf(stderr, "opt level is: %d\n",
    sess_options.config.graph_options().optimizer_options().global_jit_level());
  tensorflow::ClientSession session(root, sess_options);
  //tensorflow::graph::SetDefaultDevice("/job:localhost/replica:0/task:0/device:XLA_GPU:0", session.impl()->

  tensorflow::Status s = session.Run({{input_and_outputs.first, input_tensor}}, outputs, &output_tensors);
  s = session.Run({{input_and_outputs.first, input_tensor}}, outputs, &output_tensors);
  s = session.Run({{input_and_outputs.first, input_tensor}}, outputs, &output_tensors);

  
  struct timespec start, end;
  double total_ms = 0.0;
  std::string cur_run_num;
  for (int i = 0; i < 10; i++)
  {
    clock_gettime(CLOCK_MONOTONIC, &start);
    s = session.Run({{input_and_outputs.first, input_tensor}}, outputs, &output_tensors);
    clock_gettime(CLOCK_MONOTONIC, &end);
    double milliseconds = (end.tv_nsec - start.tv_nsec) / 1e6 + (end.tv_sec - start.tv_sec) * 1e3;
    total_ms += milliseconds;
    cur_run_num = "Run ";
    cur_run_num += std::to_string(i);
    print_timediff(cur_run_num.c_str(), start, end);
  }
  fprintf(stderr, "avg ms: %lf\n", total_ms/10);

  fprintf(stderr, "ran the graph\n");
  if (!s.ok())
  {
    fprintf(stderr, "error: %s\n", s.ToString().c_str());
    fprintf(stderr, "somethings fucked up\n");
    fprintf(stderr, "somethings fucked up\n");
    fprintf(stderr, "somethings fucked up\n");
    fprintf(stderr, "somethings fucked up\n");
    fprintf(stderr, "somethings fucked up\n");
  }

  return 0;

/*
  // This placeholder will not compile right now, but we will need it for later.
  //auto input = ops::placeholder(root.WithOpName("input"), DT_FLOAT, {-1, 56, 56, 96});

  // Matrix A = [3 2; -1 0]
  auto A = Const(root, { {3.f, 2.f}, {-1.f, 0.f} });
  // Vector b = [3 5]
  auto b = Const(root, { {3.f, 5.f} });
  // v = Ab^T
  auto v = MatMul(root.WithOpName("v"), A, b, MatMul::TransposeB(true));
  std::vector<Tensor> outputs;
  ClientSession session(root);
  // Run and fetch v
  TF_CHECK_OK(session.Run({v}, &outputs));
  // Expect outputs[0] == [19; -3]
  LOG(INFO) << outputs[0].matrix<float>();
  */
}
